vlsi/README.md

```diff
@@ -1,9 +1,72 @@
+# HAMMER VLSI flow
 This is the starting point for a vlsi flow from this repository.
 
 This flow will not work without the necessary CAD and technology plugins for HAMMER.
 
-If you are a UCB-affiliate, you may be able to acquire access to the tech-plugins.
+If you are a UCB-affiliate, you may be able to acquire access to the tool & tech plugins.
 
-# Initial Setup Instructions (For All technologies)
-Run the `init-vlsi.sh` script to pull correct versions of hammer, hammer-cad-plugins, and the hammer-tech-plugins
-```scripts/init-vlsi.sh TECH_NAME```
\ No newline at end of file
+# Initial Setup Instructions (For all technologies)
+Run the `init-vlsi.sh` script to pull correct versions of hammer, hammer-TOOL\_VENDOR-plugins, and the hammer-TECH\_NAME-plugins. Note the technologies `asap7` and `saed32` are already included and will not submodule a tech plugin.
+```shell
+./scripts/init-vlsi.sh TECH_NAME
+```
+
+An example of tool environment configuration for BWRC affiliates is given in `bwrc-env.yml`. Replace as necessary for your environment.
+
+# Example design
+## Building the design
+In this example, you will be running a SHA-3 accelerator with a dummy hard macro through the VLSI flow in the ASAP7 process. To elaborate the Sha3RocketConfig (Rocketchip w/ the accelerator) and set up all prerequisites for the build system:
+```shell
+export MACROCOMPILER_MODE=' --mode synflops'
+export CONFIG=Sha3RocketConfig
+make buildfile
+```
+Note that because the ASAP7 process does not yet have a memory compiler, synflops are elaborated instead.
+
+## Using HAMMER
+HAMMER's configuration is driven by a JSON/YAML format. For HAMMER, JSON and YAML files are equivalent - you can use either one since HAMMER will convert them to the same representation for itself.
+
+We start by pulling the HAMMER environment into the shell:
+
+```shell
+export HAMMER_HOME=$PWD/hammer
+source $HAMMER_HOME/sourceme.sh
+```
+
+The configuration for the example design is contained in `example.yml` and the entry script with hooks is contained in `example-vlsi`. You may go through Hammer's readme to learn about the supported configuration options and how to write hooks.
+
+In order to install the process, download (and optionally extract) the ASAP7 PDK tarball. Then, edit the key `vlsi.technology.asap7.tarball_dir` if you want Hammer to extract for you or `vlsi.technology.asap7.install_dir` if you have already extracted it.
+
+To synthesize the just the SHA-3 accelerator with the hard macro we have to change the physical top module (this step is not necessary if you are pushing the entire Rocket-chip through the VLSI flow):
+```shell
+export TOP=Sha3AccelwBB
```

Comment:
Is this really the best way to to the BlackBox example? Why not create a SHA3 config that uses a blackbox, and add the blackbox to the resource files in the SHA3 package?

It would be nice if `make TOP=Sha3Accel CONFIG=Sha3RocketConfig` and `make TOP=Sha3AccelwBB CONFIG=Sha3BBRocketConfig` did what they implied,

Reply:
So, a couple issues:

1. I ran into a problem when doing a `make TOP=Sha3Accel CONFIG=Sha3RocketConfig`, basically that firrtl was looking for a `firrtl_black_box_resources.f` file which didn't exist for `TOP=Sha3Accel`, so this was a workaround.

2. I can do the blackboxing in Chisel-land, but the point of doing it this way is help audiences who are using Hammer without Chisel/Firrtl. 

3. The black box we instantiate has nothing to do with the SHA-3 accelerator, so I don't think it belongs in that package.

Let me know your thoughts.

Reply:
1. Hmmm I'm not sure why that file does not exist. 

2. Can we do something like `make syn TOP_FILE=customverilog.v` for a flow which does not use Chisel/FIRRTL at all? I guess there are really three ways to use HAMMER. 1) Ingest only Chisel-generated Verilog. 2) Ingest Chisel-generated Verilog with Chisel-specified black-boxes. 3) Ingest raw verilog. 

3. I think this goes with the previous point. Should we have examples for 1 and 2, 1 and 3, or 1-3.

Reply:
1. There aren't any blackboxes in the Sha3Accel module, so I guess it doesn't get emitted (but I don't know why)

2. `customverilog.v` is already added to `extra_v_includes`, so I'd suggest having a separate `TOP` and `VLSI_TOP ?= $(TOP)` or similar which allows elaborated top to be different from the physical top module? Then all use cases can be covered by pushing through a different module.

Reply:
1. @zhemao @abejgonzalez did you deal with this stuff recently? Can you weigh in on the behavior here?

2. This makes sense to me. So if I did `make syn VLSI_TOP=my_custom_top CUSTOM_VERILOG=my_verilog.v` no Chisel/FIRRTL would be invoked?

Reply:
I guess we can introduce a `CUSTOM_VERILOG` variable that if non-empty would break all dependencies on Chisel/FIRRTL/SRAM compiler.

Reply:
1. I have never built anything without a BB... so I don't know what the behavior is for that... however that should be an easy fix.

2. IMO having Chipyard support a non-Chisel design is not the goal of this project. To me, that arbitrary vlog synthesis/par Makefile should exist in a different top level repo that we can point to in this README or the docs. We should discuss this at the Chipyard meeting. However, if we are to do this I think Jerry's idea makes sense where `CUSTOM_VLOG` is a list of Verilog files to include rather than one file.